python/ray/dag/compiled_dag_node.py

```diff
@@ -2017,62 +2046,58 @@ def _execute_until(
         execution_index: int,
         channel_index: Optional[int] = None,
         timeout: Optional[float] = None,
-    ) -> List[Any]:
-        """Repeatedly execute this DAG until the given execution index,
-        and buffer all results up to that index. If the DAG has already
-        been executed up to the given index, just return the result
-        corresponding to the given index and channel.
+    ):
+        """Repeatedly execute this DAG until the given execution index and
+        buffer results for all DagRefs that are still alive. If this comes across
+        execution indices for which the corresponding DagRef has been destructed,
+        it will release the buffer and not cache the result.
+        If the DAG has already been executed up to the given index, it will do nothing.
 
         Args:
             execution_index: The execution index to execute until.
             channel_index: The index of the output channel to get the result from.
                 Channel indexing is consistent with the order of
                 self.dag_output_channels. None means wrapping results from all output
                 channels into a single list.
-            timeout: The maximum time in seconds to wait for the result.
+            timeout: The maximum time in seconds to wait for the execution.
                 None means using default timeout (DAGContext.get_timeout),
                 0 means immediate timeout (immediate success or timeout without
                 blocking), -1 means infinite timeout (block indefinitely).
 
-        Returns:
-            The execution result corresponding to the given execution index and
-            channel index.
-
         TODO(rui): catch the case that user holds onto the CompiledDAGRefs
         """
-        if self._max_finished_execution_index < execution_index:
-            from ray.dag import DAGContext
-
-            ctx = DAGContext.get_current()
-            if timeout is None:
-                timeout = ctx.get_timeout
-
+        if not timeout:
```

Comment:
are you referring to both `0` and `None`? If you only refer to None, use `if timeout is not None`.